### PR TITLE
Improve Recent repositories settings form

### PR DESCRIPTION
**Pull request by pmiossec**
Configure repositories popup:
* Being able to see the repository selected in the lists
* Being able to do an action on multiple repositories at the same time
* Resize column size to prevent bad display (when destination list was initially empty or changing shortening strategy)
* Display 'Anchored' repositories in bold
* Add buttons to move repositories between more/less recent repositories
* Add a button to be able to remove all the deleted repositories from the list of recent repositories

Repositories list:

* Always display a separator between pinned and all recent repositories
* Add '...' to 'Configure this menu' menu entry because it will open a popup (follow windows practices)
* Add an icon in the menu for pinned repositories

Before:
image

image

After:
image

image

### CODE REVIEW

**Comment by pmiossec** [comment]
There is a last thing that I want to change but I want your point of view first...

I never really understood that notion of "Most recent repositories" and "Less recent repositories" because that's not compatible with fact that the user could modify these lists.

Shouldn't we call them something like "Preferred recent repositories" and "Other recent repositories"?

**Comment by jbialobr** [comment]
>Remove deleted repositories from the list of recent repositories

I would not do that automatically, you can add a button to trigger this action. The reason is that people can keep their repos on removal devices or subst drives. Maybe the candidates for removal should be printed in red.

**Comment by jbialobr** [comment]
>Shouldn't we call them something like "Preferred recent repositories" and "Other recent repositories"?

Hard to say as the first group is an union of preferred repos and most recent repos. 
"Preferred recent repositories" sounds to me more like intersection of preferred and recent repos.
In a case when there are no repos pinned to the most recent group, they are just most recent repositories.

**Comment by RussKie** [comment]
> Put "Ok" button on the left of the "Cancel" one

Genuinely interested why? There are so many dialogs and forms, yet I can't really see any pattern in how they have their buttons. 
For example "Process" dialog has buttons in the following order [Abort] [OK]...


> Remove deleted repositories from the list of recent repositories

Agree with @jbialobr, for the same reason I display warning icons on my dashboard PR


> Display 'Anchored' repositories in bold

What is an 'anchor' and how it is get used? And how do you anchor it?
Regardless I would probably do it by using an icon rather than making it bold. 
The same icon can then be used everywhere else - the dashboard, the repo dropdown etc - thus providing the visual continuity of experience.

> Add buttons to move repositories between more/less recent repositories

The intent is good but I find the execution uninspiring, sorry. The buttons just don't belong in that spot.
May be like this?
image
Though even that doesn't look too appealing...
The buttons should also be context aware - become enabled/disabled depending on available selection or lack of.
Also have you added the same images to the context menu?

**Comment by RussKie** [comment]
WRT: most/less recent repositories
It is something I struggle to fully comprehend as well. Is this something that really needed to be carried forward? 
Can't we maintain a list of repos in order of their access. The repo used the last sits at the top, the repo used before it sinks below it. Every time a repo is being opened/switched to - the list of recent repos gets updated...

Then the whole UI becomes simpler - in this form as well as in the dashboard, as well as a lot of code can be put to rest - reducing the tech debt and maintenance burden.
This will potentially allow for making this dialog the place to manage how repositories are presented and categorised. Maybe even re-purpose/re-use it for categories management (as part of my dashboard PR #3693)

**Comment by RussKie** [comment]
And one more thing - please squash it before merging in.

**Comment by pmiossec** [comment]
>I would not do that automatically, you can add a button to trigger this action. The reason is that people can keep their repos on removal devices or subst drives. Maybe the candidates for removal should be printed in red.

done.

>Hard to say as the first group is an union of preferred repos and most recent repos.
"Preferred recent repositories" sounds to me more like intersection of preferred and recent repos.
In a case when there are no repos pinned to the most recent group, they are just most recent repositories.

I don't really get it but I won't touch it....

>>    Put "Ok" button on the left of the "Cancel" one

>Genuinely interested why? There are so many dialogs and forms, yet I can't really see any pattern in how they have their buttons.

Because that's the Windows guidline. And yes, that should be fixed on other forms to be consistent....


>> Remove deleted repositories from the list of recent repositories

>Agree with @jbialobr, for the same reason I display warning icons on my dashboard PR

@RussKie Fixed, even before you post your very sympathic message...

>>Display 'Anchored' repositories in bold

>What is an 'anchor' and how it is get used? And how do you anchor it?
Regardless I would probably do it by using an icon rather than making it bold.

I thought about that but it's a little harder to do and because I didn't found a statisfying icon and I'm not a designer, I decided not to go with that and that my improvement will be enough for the few case this form should be used. Everyone made efforts to improve things... but I'm waiting for your PR.

>>Add buttons to move repositories between more/less recent repositories

>The intent is good but I find the execution uninspiring, sorry. The buttons just don't belong in that spot.
May be like this?

I tried also but it should require to redesign most of the form what I'm not enough confident to do. But I wait for your contribution, too....


>Though even that doesn't look too appealing...

Indeed, critics are easy....

>The buttons should also be context aware - become enabled/disabled depending on available selection or lack of.

That is already done...


>Also have you added the same images to the context menu?

No, good idea, I could do it.

>It is something I struggle to fully comprehend as well. Is this something that really needed to be carried forward?

I don't have the pretension to question choices that have been made. I just tried to improve the bad experience I felt each time I used this form (I didn't remember each time that there is a contextual menu, what is anchor, ....). This is my **simple** contribution to make it more understandable.

>And one more thing - please squash it before merging in.

No, I don't think so. Each commit is an improvement on it's own and I highly prefer that history merged instead of a big commit. But the maintainer/merger will choose what he prefers now that it is easy to do thanks to github....

**Comment by EbenZhang** [comment]
> No, I don't think so. Each commit is an improvement on it's own and I highly prefer that history merged instead of a big commit. But the maintainer/merger will choose what he prefers now that it is easy to do thanks to github....

Agree, as long as the commit is logic. 
It's hard to know the reason of the change in a large commit when investigating a bug.

**Comment by jbialobr** [comment]
>Can't we maintain a list of repos in order of their access. The repo used the last sits at the top, the repo used before it sinks below it. Every time a repo is being opened/switched to - the list of recent repos gets updated...

I don't like when items in the menu constantly change theirs position. 

>Then the whole UI becomes simpler - in this form as well as in the dashboard, as well as a lot of code can be put to rest - reducing the tech debt and maintenance burden.
This will potentially allow for making this dialog the place to manage how repositories are presented and categorised. Maybe even re-purpose/re-use it for categories management (as part of my dashboard PR #3693)

I think this menu should share the same display logic as the dashboard proposed in #3693.

**Comment by jbialobr** [comment]
>>And one more thing - please squash it before merging in.

>No, I don't think so. Each commit is an improvement on it's own and I highly prefer that history merged instead of a big commit. But the maintainer/merger will choose what he prefers now that it is easy to do thanks to github....

I reacted with thumb up to the @EbenZhang comment, but I want to state it explicitly:
The more details we know about the reason standing behind a commit the easier is to make a judgement if the change is correctly implemented. Keeping commits small and atomic helps to achieve that state.
So, many small commits are welcome.

**Comment by RussKie** [comment]
> So, many small commits are welcome.

I was going to suggest the same.

**Comment by RussKie** [comment]
> I just tried to improve the bad experience I felt each time I used this form

This is great! 

> I don't have the pretension to question choices that have been made. 

...And this isn't great. 




>  ...but I'm waiting for your PR.
> But I wait for your contribution, too....

Unless I am misreading it, you expect me to come and fix things after you? I find it bewildering... 


> Indeed, critics are easy....

There is a distinction between "criticism" and "feedback". The former is to say something is bad, the latter is to say something isn't right and suggest a possible alternative.

No need to feel defensive when there is feedback. UIs are hard, and sometimes they take a lot of effort and iterations to get right, for them to feel natural, easy to use and look appeasing. 
It is ok to make an effort and fail - this way we learn 😃 ...as long as we stand up and keep on walking.

How about left-aligning the buttons like this:
image

**Comment by RussKie** [comment]
> I don't like when items in the menu constantly change theirs position.

The repo which you're currently working with will be at the top... But that was a random idea anyway not a formed proposal.

> I think this menu should share the same display logic as the dashboard proposed in #3693.

Totally! Are you/we open to reconsidering the current implementations?
The UX should be unified between repo history, categories, sort/grouping order and the navigation. Right now it is very fragmented and disjoint.


What is an 'anchor' and what does it get used for anyway?

**Comment by jbialobr** [comment]
>Are you/we open to reconsidering the current implementations?

I would like to. I think that repository dropdown could be populated with repositories grouped in categories - the same categories as on dashboard, with uncategorised including and preserving the order of categories and repos set for the dashboard. For me the most recent repos group acts as a 'favorites' category with a constant set of pinned repos.
 
An anchor is used to pin a repo to most/less recent repositories group.

File: GitUI/CommandsDialogs/BrowseDialog/FormRecentReposSettings.cs
Comment (jbialobr) [inline GitUI/CommandsDialogs/BrowseDialog/FormRecentReposSettings.cs]: Can you make it non-static and derive from the actual list font? Is it a performance optimization?

**Review by jbialobr — DISMISSED** [review]

**Comment by GintasS** [comment]
This is an old PR and should be closed or set to a draft. @RussKie

**Comment by pmiossec** [comment]
I don't go to this screen very often but when I do it to manage the pinned repositories, I find it horrible.
So I took the time to refresh (5 years later!) this PR.

For me, it improves a lot the UX and if it could be reviewed again. I have updated the screenshot.

For me, this idea of "anchor" in "Pinned" or "Recent" repositories is difficult to understand and don't bring a lot of value. And I'm pretty sure no one use it. **Are you OK if I remove it?**

**Comment by RussKie** [comment]
> For me, this idea of "anchor" in "Pinned" or "Recent" repositories is difficult to understand and don't bring a lot of value. And I'm pretty sure no one use it. **Are you OK if I remove it?** To be clearer, I will only keep the idea to pin a repository but not anchor a repository in "recent" list.

💯 

If my memory serves me, I asked about those quite a few times in the past.

Also, I think we should move this dialog into the Settings, so it's more discoverable and accessible.

**Comment by mstv** [comment]
Looks useful.

> Add an icon in the menu for pinned repositories

"Pinning" might work if one uses the Dashboard only or if the "Maximum number of pinned recent repositories" is 0. In the second case, this label is ... misleading.
This list is still automatically updated with the most recently opened repos - at least those chosen from "this" menu as well as opened via the CLI, including the shell extension.

**Comment by pmiossec** [comment]
> > For me, this idea of "anchor" in "Pinned" or "Recent" repositories is difficult to understand and don't bring a lot of value. And I'm pretty sure no one use it. **Are you OK if I remove it?** To be clearer, I will only keep the idea to pin a repository but not anchor a repository in "recent" list.
> 
> 💯
> 
> If my memory serves me, I asked about those quite a few times in the past.

Done removed "anchored" notion

> Also, I think we should move this dialog into the Settings, so it's more discoverable and accessible.

This PR is becoming big and I don't know the difficulties ahead so I think it's better to do it in a following PR...

@mstv Can you be more explicit? I don't really get you.

>"Pinning" might work if one uses the Dashboard only or if the "Maximum number of pinned recent repositories" is 0. In the second case, this label is ... misleading.

Not sure to get the idea. In fact I don't use the Dashboard because I don't manage a lot of repositories in a day to day workflow (less than 5) and in this case, using the dashboard is a lot less handy. So I only use the repos switcher in the toolbar.
You mean that "pinned" repos are not displayed in the dashboard and should be added? Or you have another idea in mind?

Also for me, the "Maximum number of pinned recent repositories" setting should be removed and we must always display all the pinned repositories.

>This list is still automatically updated with the most recently opened repos - at least those chosen from "this" menu as well as opened via the CLI, including the shell extension.

Yes that's the "all recent" list and I find that normal. I have added the separator so that the distinction is clearer

**Comment by RussKie** [comment]
Are the screenshots up the top up-to-date?

> Also for me, the "Maximum number of pinned recent repositories" setting should be removed and we must always display all the pinned repositories.

There should be a limit on how many pinned repositories you can have. Hard or soft (i.e., configurable).

File: GitUI/CommandsDialogs/BrowseDialog/FormRecentReposSettings.cs
Comment (mstv) [inline GitUI/CommandsDialogs/BrowseDialog/FormRecentReposSettings.cs:130]: Use explicit types, please. (`Ctrl+.` gives support)

**Comment by mstv** [comment]
https://github.com/gitextensions/gitextensions/pull/10084#issuecomment-1229267521 still is valid.

With the settings I use, non-zero "pinned" recent repos, this has nothing to do with "pinned", "anchored" or another interpretation - it is simply "the most recently opened repos".
So the label of this setting should be restored: `Maximum number of most recent repositories`.
Only in case this setting value is zero, the upper right list should be labeled "pinned". (Currently, I am unable to review how this re-use of this list as "pinned repos" is implemented.)
If the the setting is non-zero we should not falsely represent the user could pin some repos. I.e. the "pin" buttons should be disabled.

image


>> This list is still automatically updated with the most recently opened repos - at least those chosen from "this" menu as well as opened via the CLI, including the shell extension.
>
> Yes that's the "all recent" list and I find that normal. I have added the separator so that the distinction is clearer

No, both lists are updated automatically. User pinning gets lost.

And no, there is no need to add a separator - at least in case of the non-zero setting:

image

**Comment by gerhardol** [comment]
Slightly ot, very unlikely that i will be lengthy since I am away from any computer and a sharp mind for unknown time

I regret I agreed to change from _most recent_ to _pinned_ as this makes the functionality to complicated even for me.
But anchored could be removed

A list of 50 recent is too short for me.
I have opened a feature a long time ago to not add submodules to recents. Not implemented previously as I found it too special. But it is what would fit me, in _less recents_ only. Then there is a LRU for superrepos too.

**Comment by gerhardol** [comment]
@pmiossec rebase this on master or target 4.1?

I do not want to delay 4.1 any more...

**Comment by mstv** [comment]
> I do not want to delay 4.1 any more...

👍

**Comment by pmiossec** [comment]
> @pmiossec rebase this on master or target 4.1?

I have removed all the commits that raised the discussion to keep only the little improvement that makes using this form a little more easier. 
And I would like to see a start of improvement included in v4.1
I don't go very often in this screen but every time I do, the experience is really bad... 

The code is finished and the branch is in good state but I had to go and didn't had the time to update the PR. I will do tomorrow morning... 

After that, you will whatever you want with it hoping it won't delay the release...

**Pull request closed without merging**
